framework/include/transfers/MultiAppGeneralFieldTransfer.h

```diff
@@ -0,0 +1,590 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MultiAppConservativeTransfer.h"
+#include "MooseHashing.h"
+#include "KDTree.h"
+
+#include "libmesh/generic_projector.h"
+#include "libmesh/meshfree_interpolation.h"
+#include "libmesh/system.h"
+#include "libmesh/mesh_function.h"
+#include "libmesh/parallel_algebra.h" // for communicator send and receive stuff
+
+/**
+ * It is a general field transfer. It will do the following things
+ * 1) From part of source domain to part of domain. Support subdomains/boundaries to
+ *    subdomains/boundaries, mixing as appropriate
+ * 2) interpolation and extrapolation, as appropriate
+ * 3) Support higher order FEM
+ * 4) Support mixed orders between source and target variables
+ * 5) Support both distributed and replicated meshes
+ * 6) Support both origin and target displaced meshes
+ * 7) Support siblings transfers
+ * 8) Support multiple child apps in both the transfer source and target
+ */
+class MultiAppGeneralFieldTransfer : public MultiAppConservativeTransfer
+{
+public:
+  static InputParameters validParams();
+
+  MultiAppGeneralFieldTransfer(const InputParameters & parameters);
+
+  virtual void initialSetup() override;
+
+  virtual void execute() override;
+
+  /// Get the source variable name, with the suffix for array/vector variables
+  VariableName getFromVarName(unsigned int var_index);
+
+  /// Get the target variable name, with the suffix for array/vector variables
+  VariableName getToVarName(unsigned int var_index);
+
+  // This needs to be moved into libMesh according to Fande
+  // The hash helps sort the points in buckets
+  // Hashing the floats also leads to floating point comparison errors in their own way
+  // We hash a rounding of the float instead
```

Comment:
I ll see if I can make compare_point and hash_point more compatible. 

I dont really want to switch to a KD Tree, for a few reasons:
- the reconstruction of the solution (in its entirely) is already the expensive part in the profiles for higher order elemental variables. The KDTree wont beat the map
- it s a significant rework because the map is built progressively through the transfer, a KDTree would need to be built at the end. Likely we would keep the map (or a vector) and build the KDTree with it.
- 1e-10 even meters (because why would you use a km-scale mesh to look at mm-scale phenomena) is quite below anything anyone does in MOOSE. I can change it to 1e-17 for sure. 
A variable scale (maybe using the bounding box of the app) could work too but I m not sure how to make it enter these compare_point/hash_point routines.

Reply:
btw all we need here is that `point` is close enough to `coordinateTransform^-1(coordinateTransform(point))`
The map is filled with one and retrieved with the other.

Reply:
another thing is that we specifically do not want a Nearest Node type behavior. If the point is not in the map, we want to resort to use the current value of the variable, not the nearest. Or at least that's how Fande wrote this retrieval of variable values. We could do either of course. 

Of course the distance from the point in a KDTree being available, we could reproduce the previous behavior by examining the distance and rejecting the value if it's beyond an arbitrary (uh oh) size

Reply:
umm I see your point now.
I m actually more worried about 1000000.000000001 now than than 1e-10. The former will force a very strong agreement between the values : all the decimals before 0 + 10 after 0 will get excessive. 

I think I have a solution though, we ll see

Reply:
@roystgnr what do you think of the new heuristic I pushed?

Reply:
I'll love to see the solution.  We don't want to get just any nearest node, but we do want Nearest-Node-iff-within-tolerance behavior, and there's no way to find every match within tolerance with a single hash lookup.  To get O(1) instead of O(N) requires having more than one bucket in the hash table, and then any hash(Point) is a function from R^d to a disconnected space, so it's discontinuous; there will always exist some p for which hash(p) and hash(p+epsilon) are going to end up in different buckets, no matter what hash we choose and no matter what size we make epsilon.  After that, it doesn't matter what our equality function is, because `unordered_map` isn't going to be testing a point in one bucket for equality with a point in a different bucket.

We could force such extra tests, I guess?  If we do the `round()` trick to spatially-widen our hash buckets, then we could also see how close we are to a bucket face where we'd have rounded differently, and if we're too close then we perturb p to put it in the other bucket and we query two buckets before declaring that something is or isn't in the `unordered_map`.  (or we query 4 buckets if we're near an edge, or 8 near a vertex...)  That would be faster than a KD Tree at least.  We'd want to wrap the `unordered_map` in some other class to make sure we don't accidentally use "raw" access instead in the future.

Ooh, email tells me I'm out of date here.  I'll check out your new heuristic now.

Reply:
I think it's falling into the same insurmountable problem.  I tossed a simplified version (tell me if I screwed something up!) into a utility to play with:

```
#include <cmath>
#include <iomanip>
#include <iostream>

double getMantissaRoundedPoint(const double p)
{ 
  // - get the scientific notation for the number
  // - use round on the mantissa * 1e12
  // - if rounding to the next power of 10, adjust the exponent
  // - form a new point using the scientific notation
  int exponent = (std::fabs(p)<1e-10) ? 0 : (int)std::log10(std::fabs(p));
  double rounded_m = std::round(p / pow(10, exponent - 12));
    
  if (std::fabs(rounded_m) == 1e13)
    {
      exponent += 1;
      rounded_m = (rounded_m > 0) ? 1e12 : -1e12;
    }  
  
  return rounded_m * pow(10, exponent - 12);
}

int main(int argc, char ** argv)
{ 
  if (argc < 2)
    {
      std::cout << "Needs 2 arguments" << std::endl;
      return 1; 
    }  
    
  const double p = std::atof(argv[1]);
  
  std::cout << std::setprecision(17);
  std::cout << "Rounded of " << p << " is " << getMantissaRoundedPoint(p);
  
  return 0;
} 
```

And playing around randomly it didn't take long to find:
```
> g++ test.C -o test.x && ./test.x 1.2345678901235
Rounded of 1.2345678901235 is 1.234567890124
> g++ test.C -o test.x && ./test.x 1.234567890123499
Rounded of 1.2345678901234991 is 1.2345678901229999
```

One point is perturbed less than 9e-16 from the other, but they will get hashed into different buckets.  I'm becoming increasingly sure that either we need to rework the outer code to avoid ever needing to insert a point and query it from a perturbed point, or we need to be prepared to test multiple neighboring buckets for every query near a hash boundary.

(Or we could make the hash mesh-dependent in some way? Move its discontinuities to places where they're not within `TOLERANCE` of any possible existing quadrature point locations?  That sounds much harder than any option save for "find a non-constant continuous function from R to N", though.)

Reply:
I think we d need to test at least 8 extra bins (combinations of slight errors either up or down). That doesn’t seem desirable. 

This heuristic solves the issue of being valid for both 1e-10 and 1e10.

It indeed does not solve the rounding boundaries at 5e-13. The point you tested is right on a bucket boundary, but it would be equivalent for a lot of points that would be near it.

I ll see if I can think of something, otherwise KDtree it will be

Reply:
We would need to test up to 7 extra bins, but most of the time we'd be testing just the 1 original and 0 extra.  I think it'd be faster than a KDTree, though it might also be more work.

I'd be happy with a KDTree, though; "can break silently" terrifies me, but "should run a little faster" we can fix later.

Reply:
Another solution to this would be to stop indexing by float, and start numbering quadrature points instead. We would need to rework the projector, but we could ditch the map for a vector and make the searches much easier. 
I dont know if we have a system for numbering quadrature points?

Reply:
Another solution:
- use a lot less bins (eg wider bins). We realistically do not need more than 6 digits  per coordinate to describe a unique point in the mesh with scientific notation (already roughly 1e18 bins in 3D).

But the chance of being on a bin edge (where floating point calcs can make us take one side or another) with 6 digits are 1e6 less than with 12 digits.

We also have the following safeguards:
- if double insertion, we have a conflict detection already in place, it is detected and output as a warning to the user
- if no value found (retrieval rounding different than insertion rounding), we can keep track of it and output a warning (error if error_on_miss is true). I m yet to add that one

Reply:
> * use a lot less bins (eg wider bins). We realistically do not need more than 6 digits  per coordinate to describe a unique point in the mesh with scientific notation (already roughly 1e18 bins in 3D).

But with that, you would still be able to engineer a number to break it, like you did for 12 digits.
Basically it will all be a tradeoff between the number of digits in the scientific notation, somewhat correlated to the number of (non-zero at the end) digits in the floating point representation. More digits used for the bins make them less human-likely, but more likely to be generated by the machine out of FP operations. What a mess haha

Reply:
> so 27 bins total we could have to check

The trick is to narrow it down by *which* hash boundary we're near.  Worst case scenario is we're near a hash bin vertex, and then we have 2^d other bins to check; only if we weren't able to tell which vertex would we need to check 3^d.

> A final option, which I dont think i ll favor, would be to insert into the map multiple times if we are near a hash boundary. This would ensure that searches are always successful.

I thought about this one; it would be an improvement in the Write-Once-Read-Many case, but I don't think it would make a difference in our usage.

> stop indexing by float, and start numbering quadrature points instead

A quadrature point on the querying side can be an arbitrary point on the response side, so a system like (element id, quadrature point id) wouldn't really be meaningful to it, so we'd have to send that as a label from the querying side ... but if we can keep track of the labeling we could just as well use the queried `Point` as the label.

> use a lot less bins (eg wider bins).

Fewer bins means fewer boundaries means lower odds of crossing one, yeah, but we ought to be able to get our failure rate down to 0 here, not just some 1e-6*C.  More annoyingly: unless we treat our `Point` values as fixed-precision arithmetic and offset the bins in some weird way, the kinds of "crossed to the wrong side of the bin boundary" numbers that you can find everywhere via deliberate engineering can be found in disproportionate surplus from some kinds of mesh generators.  Cubit's options for curved (and even non-curved, when using NURBS) meshes have a lovely tendency to take numbers like 0.0 and 1.0 (prime bin boundary values) and do a +/- epsilon on them when reconstructing vertex coordinates from spline node coordinates.

Let me know if you feel like you're banging your head against the wall on this one and I can take a crack at the "check 2^d bins" option.  (I can feel "if we can keep track of the labeling we could just as well use the queried `Point` as the label" calling to me too, but that first "if" scares me.)

Reply:
> A quadrature point on the querying side can be an arbitrary point on the response side, so a system like (element id, quadrature point id) wouldn't really be meaningful to it, so we'd have to send that as a label from the querying side ... but if we can keep track of the labeling we could just as well use the queried Point as the label.

The response side only needs the point, not whether it s a quadrature point and where it is exactly. The index of the quadrature point, if this existed, would entirely stay on the querying side.

Reply:
Numbering quadrature point would be quite a clean up as well, we could rely on the same map < dof_id_type , pair<value, distance, problem> > instead of having these Point maps only for higher order elemental variables, which even though they remove the storage of the problem id, they cost more because of the size of a Point.

The problem is how to do it, given that there are multiple problems which may not have the same number of elements or Qps per element. 

Anyhow, I implemented the multi-search point map.

I had something that seemed to by comparing to the boundaries, but to be honest it was messy, it depended on 3 different epsilons and it wasnt reliable. 
In the last commit I went to the 27 point search and it seems to work reliably. It might not be the fastest.

Reply:
The current map is working for every case I can come up with now.